ROCANA-9231: Fork PerFieldPostingsFormat

bradcupit opened the pull request.

joey commented on src/main/java/com/rocana/lucene/codec/v1/RocanaPerFieldPostingsFormat.java:
```diff
@@ -48,35 +50,63 @@
+ *   - The original code called {@link PostingsFormat#forName(String)}
+ *     which caused a problem. Since our fork of Lucene's postings format,
+ *     {@link RocanaLucene50PostingsFormat}, does not have a custom short name
+ *     and we didn't register it via SPI, then {@link PostingsFormat#forName(String)}
```
The issue isn't that we can't change the name or register it in the SPI, it's that we need to change the behavior of the `Lucene50PostingsFormat` that is already registered and written on disk.

joey commented on src/main/java/com/rocana/lucene/codec/v1/RocanaPerFieldPostingsFormat.java:
```diff
@@ -48,35 +50,63 @@
+ *     short name, when we see that short name we instead return Rocana's fork of the
+ *     Lucene postings format.
+ *   - Removed trailing whitespace.
+ *   - Changed these javadocs.
```
You may also want to mention the rename and new package.

mpeterson0x0 commented on src/main/java/com/rocana/lucene/codec/v1/RocanaPerFieldPostingsFormat.java:
```diff
@@ -274,6 +304,30 @@ public FieldsReader(final SegmentReadState readState) throws IOException {
+     */
+    private PostingsFormat lookupPostingsFormat(String postingsFormatShortName) {
+      if (postingsFormatShortName.equals("Lucene50")) {
+        RocanaSearchCodecV1 codec = (RocanaSearchCodecV1) Codec.forName(RocanaSearchCodecV1.NAME);
```
There is a loading cost with no benefit to calling `Codec.forName` and then casting to a concrete type.  You should either just instantiate the concrete type directly (faster) or create an interface `RocanaSearchCodec` with the `getActualPostingsFormat` (extensible).

joey replied:
The loader caches a single instance of the codec, so I'm sure that instantiating a new instance each time would be faster.

bradcupit replied:
@mpeterson0x0 I wanted to get "the one and only instance of our postings format" which we could do by looking it up in SPI, but since we don't have store our postings format in SPI (yet) it seemed like a good alternative to get the one and only instance of our codec and then use that to get the `final` postings format field.

joey commented on src/main/java/com/rocana/lucene/codec/v1/RocanaPerFieldPostingsFormat.java:
```diff
@@ -274,6 +304,30 @@ public FieldsReader(final SegmentReadState readState) throws IOException {
+     *
+     * The original Lucene code did an SPI (Service Provider Interface) lookup to
+     * get the postings format. The problem is we forked the postings format, but
+     * didn't register it via SPI -- for the reason see:
```
You probably need to update this comment following our discussion this morning.

joey commented on src/main/java/com/rocana/lucene/codec/v1/RocanaSearchCodecV1.java:
```diff
@@ -126,7 +126,38 @@ Codec getWrappedCodec() {
+   *
+   * Typical Lucene code would look up the postings format via SPI (Java's
+   * Service Provider Interface), but we don't give our forked postings
+   * format a custom short name, nor do we register it via SPI (see
```
Needs to be updated per our discussion.

joey reviewed (changes requested):
The code looks solid @bradcupit! I think some of the comments need to be modified per our discussion this morning.

Commit:
ROCANA-9231: Fork PerFieldPostingsFormat (part 1 of 2)

Commit:
ROCANA-9231: Fork PerFieldPostingsFormat (part 2 of 2)

Add javadocs and remove trailing whitespace

This must be a separate commit, not squashed. If we squash it the changes to TestRocanaPerFieldPostingsFormat.java will be too much and git will think we've deleted + added a new file rather than renamed and modified a file. That means we won't be able to do a diff from the current version of this file to the very first version committed (which is the pristine code, unmodified, from Lucene). That diff is valuable as it shows precisely what's changed in our fork.

Unfortunately git uses heuristics to determine when a file's been renamed and there's so many changes to this file it breaks that algorithm. For more details: https://git.wiki.kernel.org/index.php/GitFaq#Why_does_Git_not_.22track.22_renames.3F

bradcupit commented on src/test/java/com/rocana/lucene/codec/v1/TestRocanaPerFieldPostingsFormat.java:
```diff
@@ -0,0 +1,64 @@
+/*
```
This is not a new file, it's a rename + modification of TestPerFieldPostingsFormat, which github is showing as a deleted file.

In the git history, however, they're shown as the same file (though, I have to jump through a hoop to make it happen, see [this commit message](https://github.com/scalingdata/rocana-lucene-codec/pull/10/commits/dc84f7665c43f3000607422d33391aab70f85f02)).

mpeterson0x0 commented:
My point was that this code is not extensible to future versions. If you
going to pay the cost of a dynamic lookup that uses reflection to create a
instance of a particular concrete impl (as denoted by your cast), then
there's no point in a dynamic lookup - you may as well avoid the cost of
reflection based instantiation. If however you want to make this extensible
to get future versions of the RocanaSeachCodec (e.g., V2), then you should
create an interface "RocanaSearchCodec" interface that has the
"getActualPostingsFormat" method on it and cast to that in this method.

On Fri, Jan 27, 2017 at 1:19 PM, Brad Cupit <notifications@github.com>
wrote:

> *@bradcupit* commented on this pull request.
> ------------------------------
>
> In src/main/java/com/rocana/lucene/codec/v1/RocanaPerFieldPostingsFormat.
> java <https://github.com/scalingdata/rocana-lucene-codec/pull/10>:
>
> > +     *
> +     * The original Lucene code did an SPI (Service Provider Interface) lookup to
> +     * get the postings format. The problem is we forked the postings format, but
> +     * didn't register it via SPI -- for the reason see:
> +     * {@link RocanaSearchCodecV1#getActualPostingsFormat()}
> +     *
> +     * This meant our custom codec was using our forked postings format, but here
> +     * the SPI lookup returned Lucene's original postings format. So we're mixing
> +     * postings formats.
> +     *
> +     * In this method we'll catch those lookup calls for Lucene's postings format
> +     * ("Lucene50") and instead return Rocana's fork.
> +     */
> +    private PostingsFormat lookupPostingsFormat(String postingsFormatShortName) {
> +      if (postingsFormatShortName.equals("Lucene50")) {
> +        RocanaSearchCodecV1 codec = (RocanaSearchCodecV1) Codec.forName(RocanaSearchCodecV1.NAME);
>
> @mpeterson0x0 <https://github.com/mpeterson0x0> I wanted to get "the one
> and only instance of our postings format" which we could do by looking it
> up in SPI, but since we don't have store our postings format in SPI (yet)
> it seemed like a good alternative to get the one and only instance of our
> codec and then use that to get the final postings format field.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scalingdata/rocana-lucene-codec/pull/10>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/ANP6-7yQU4hcqOKHhP_I7U_gatMQK2nMks5rWjVPgaJpZM4Lv60C>
> .
>



-- 
Thank you,
Michael Peterson

Infrastructure Software Engineer
919-627-4640
@quux00

bradcupit commented:
@mpeterson0x0 I wonder if we're having a miscommunication? I'll try and address your concerns individually, but if you feel like i'm missing the point then let's jump on a zoom.

> If you going to pay the cost of a dynamic lookup that uses reflection to create a instance of a particular concrete impl

The Lucene code only uses reflection to create the impl once. After that it's cached.

> then there's no point in a dynamic lookup

The purpose of the dynamic lookup was not because I wanted a dynamic lookup, but to get the "singleton" instance of the codec. I don't mean singleton like GoF singleton, rather, "the one instance of the codec Lucene instantiated".

From there I could get the postings format, and then we know we have the same postings format the codec uses (not a separate instance).

> If however you want to make this extensible to get future versions of the RocanaSeachCodec (e.g., V2), then you should create an interface "RocanaSearchCodec" interface that has the "getActualPostingsFormat" method on it and cast to that in this method.

If/when we need a V2 codec we'd need the ability to use both codecs at once: one for new data and one for historical data. At that point we'd need a new short name ("RocanaSearchCodecV1" and "RocanaSearchCodecV2"). We'd need to specify just one of those for the dynamic lookup so I don't think the interface would buy us anything. However, if it did, we could certainly add the interface at that time. I like the idea of adding the interface later when we know we need it.

mpeterson0x0 commented:
Right, sorry. I forgot that the codec SPI name changes between versions.
That seems unnatural and as if it partially defeats the purpose of a
"dynamic lookup", so not totally sure why Lucene does it that way.

Thanks for taking the time to clarify.

-Michael

On Mon, Jan 30, 2017 at 11:23 AM, Brad Cupit <notifications@github.com>
wrote:

> @mpeterson0x0 <https://github.com/mpeterson0x0> I wonder if we're having
> a miscommunication? I'll try and address your concerns individually, but if
> you feel like i'm missing the point then let's jump on a zoom.
>
> If you going to pay the cost of a dynamic lookup that uses reflection to
> create a instance of a particular concrete impl
>
> The Lucene code only uses reflection to create the impl once. After that
> it's cached.
>
> then there's no point in a dynamic lookup
>
> The purpose of the dynamic lookup was not because I wanted a dynamic
> lookup, but to get the "singleton" instance of the codec. I don't mean
> singleton like GoF singleton, rather, "the one instance of the codec Lucene
> instantiated".
>
> From there I could get the postings format, and then we know we have the
> same postings format the codec uses (not a separate instance).
>
> If however you want to make this extensible to get future versions of the
> RocanaSeachCodec (e.g., V2), then you should create an interface
> "RocanaSearchCodec" interface that has the "getActualPostingsFormat" method
> on it and cast to that in this method.
>
> If/when we need a V2 codec we'd need the ability to use both codecs at
> once: one for new data and one for historical data. At that point we'd need
> a new short name ("RocanaSearchCodecV1" and "RocanaSearchCodecV2"). We'd
> need to specify just one of those for the dynamic lookup so I don't think
> the interface would buy us anything. However, if it did, we could certainly
> add the interface at that time. I like the idea of adding the interface
> later when we know we need it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scalingdata/rocana-lucene-codec/pull/10#issuecomment-276109202>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANP6-75oM-uj-yGVyH-a0IidfRQPl7azks5rXg6dgaJpZM4Lv60C>
> .
>



-- 
Thank you,
Michael Peterson

Infrastructure Software Engineer
919-627-4640
@quux00

joey commented on src/test/java/com/rocana/lucene/codec/v1/RocanaBasePostingsFormatTestCase.java:
```diff
@@ -48,13 +48,37 @@
+ *
+ * Why we forked:
+ *   - Some code uses Lucene's {@link PerFieldPostingsFormat} but we need it to
+ *     also recognize our fork of that class: {@link PerFieldPostingsFormat}.
```
Should this link point to `RocanaPerFieldPostingsFormat`?

joey reviewed (approved):
I found one possible typo in a javadoc, but otherwise this is ready to merge. Thanks @bradcupit!
